Migrated the Xtend code from 'org.eclipse.xtext.xtext.ui' project to Java

nbhusare opened the pull request:
- [x] Migrated Xtend code to Java 
- [x] Ran the Maven/Gradle build locally
- [x] Removed the "org.eclipse.xtend.lib" plug-in dependency from the manifest file
- [x] Adapted build.properties file
- [x] Adapted .classpath file

Signed-off-by: nbhusare <neerajbhusare@gmail.com>

nbhusare commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/AdvancedNewProjectPage.java:
```diff
@@ -0,0 +1,489 @@
+	}
+
+	@Override
+	public void createControl(Composite parent) {
```
@cdietrich @szarnekow @tivervac The size of this method and the nested lambda expressions made the task of translating to Java complicated. After migrating, I have verified this method twice; however, I may have overlooked something. I would appreciate it if you could do a thorough review of this method.

tivervac replied:
LGTM except 1 comment. I feel your pain.

nbhusare commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/AdvancedNewProjectPage.java:
```diff
@@ -0,0 +1,489 @@
+		setPageComplete(statusWidget.getSevertity() != IMessageProvider.ERROR);
+	}
+
+	public void checkWidgets(SelectionEvent e) {
```
@cdietrich @szarnekow @tivervac Thorough review needed for this method too.

nbhusare replied:
I have verified the code changes by launching the wizard and going through all the usecases.

nbhusare commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/AdvancedNewProjectPage.java:
```diff
@@ -0,0 +1,489 @@
+		if (autoSelectIdeProject) {
+			autoSelectIdeProject = false;
+			String message = "'" + createIdeProject.getText() + "' project was automatically selected as option '"
+					+ ((Button) source).getText() + "' requires it.";
```
@cdietrich @szarnekow @tivervac The value of `source` can possible be `null` here. It is references at multiple points in this code, however, it might fail here. I can add safety check code, however, it is not clear to me where to add it. Please suggest.

nbhusare replied:
After a bit of debugging, I learned that the control would come in the current if block and the following one only if the value of `autoSelectIdeProject` and `autoSelectSDKProject` is true. For this to happen, you have to select either “Eclipse Plug-in” or “Create update site” option, in which case the value of the `source` object won't be null.

I think it is ok if we don't add any safety check for the `source` object.

nbhusare commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/AdvancedNewProjectPage.java:
```diff
@@ -0,0 +1,489 @@
+		createSDKProject.setSelection(false);
+		createP2Project.setSelection(false);
+
+		select(preferredBuildSystem, BuildSystem.NONE);
```
The original code gets the enum using `BuildSystem.values.head`. I have replaced it with `BuildSystem.NONE`. Similar change is done in the below two lines.

tivervac replied:
LGTM

tivervac commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/ui/editor/hierarchy/XtextCallHierarchyLabelProvider.java:
```diff
@@ -0,0 +1,108 @@
+
+		List<QualifiedName> qualifiedNames = Arrays.asList(hierarchyNode.getElement().getQualifiedName(),
+				hierarchyNode.getGrammarDescription() != null ? hierarchyNode.getGrammarDescription().getQualifiedName() : null);
+		return Joiner.on(" - ").join(Iterables.filter(qualifiedNames, (qualifiedName) -> qualifiedName != null));
```
You can use Objects::nonNull here

tivervac commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/StatusWidget.java:
```diff
@@ -0,0 +1,122 @@
+			status = setStatus(severity, text, () -> {}, () -> {});
+		} else {
+			link.setText(link.getText() + "\n" + text);
+			Matcher matcher = Pattern.compile("<a>(.*)</a>").matcher(text.trim());
```
Precompiled regex can be reused here

tivervac commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/StatusWidget.java:
```diff
@@ -0,0 +1,122 @@
+		setVisible(severity != IMessageProvider.NONE);
+		imageLabel.setImage(imageFor(severity));
+		link.setText(text);
+		Matcher matcher = Pattern.compile("<a>(.*)</a>").matcher(text.trim());
```
Not sure how often this has to run but precompiling the regex might be a good idea

tivervac commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/AdvancedNewProjectPage.java:
```diff
@@ -0,0 +1,489 @@
+						createIdeProject.setEnabled(true);
+					}
+					// auto deselect SDK and p2 project when Eclipse plug-in is deselected
+					if (e.getSource() == createUiProject) {
```
Isn't this `.equals`?

nbhusare replied:
No. It is identity comparison.

tivervac commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/AdvancedNewProjectPage.java:
```diff
@@ -0,0 +1,489 @@
+				reportIssue(ERROR, message, () -> createWebProject.setSelection(false));
+			} else {
+				String message = "To build the '" + createWebProject.getText() + "' " //
+						+ "project, you need to choose Maven or Gradle build system." //
```
build system -> as build system

or

choose Maven -> choose the Maven

tivervac commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/AdvancedNewProjectPage.java:
```diff
@@ -0,0 +1,489 @@
+	}
+
+	protected void select(Combo combo, Enum<?> enu) {
+		String[] items = combo.getItems();
```
Did you test this method? It's different than in Xtend. Don't have an IDE handy to tell me if it's correct

nbhusare replied:
Yes, I have tested this method.

tivervac commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/AdvancedNewProjectPage.java:
```diff
@@ -0,0 +1,489 @@
+		} else if (junitVersion5.getSelection()) {
+			return JUnitVersion.JUNIT_5;
+		} else {
+			throw new IllegalStateException();
```
Might be nice to add a message to this exception e.g. "JUnit version could not be determined"

nbhusare commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/AdvancedNewProjectPage.java:
```diff
@@ -0,0 +1,489 @@
+		Object source = null;
+		if (e != null) {
+			source = e.getSource();
+		}
```
The method `CheckWidgets()` is called with `null` when we select the link (for example, [Gradle](Gradle) in the image below) in the proposed error message. If we add an else statement, and report/add an issue here, the finish button of the wizard will remain disabled, even if all the conditions in the if blocks below are met. Given the situation, I'll keep the code AIS.

![image](https://user-images.githubusercontent.com/3328139/77845852-90d0e680-71cf-11ea-8458-c9cca3002125.png)

nbhusare commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/StatusWidget.java:
```diff
@@ -0,0 +1,124 @@
+		imageLabel = new Label(this, SWT.NONE);
+		imageLabel.setText("   ");
+		GridData imgLabelLayoutData = new GridData(GridData.VERTICAL_ALIGN_BEGINNING);
+		imgLabelLayoutData.widthHint = 20;
```
PS image below. Changed the value of `widthHint` from 15 to 20 pixels. There were cases where I found that the image was getting cropped. Changing it to 20 pixes resolves that issue, also the additional pixels improves the user experience.

![image](https://user-images.githubusercontent.com/3328139/77846757-e3150600-71d5-11ea-82c0-085a2d03090a.png)

nbhusare commented on org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/wizard/project/StatusWidget.java:
```diff
@@ -0,0 +1,124 @@
+
+		link = new Link(this, SWT.NONE);
+		GridData linkLayoutData = new GridData(GridData.FILL_HORIZONTAL);
+		linkLayoutData.heightHint = 40;
```
PS images below.

**Before**: `linkLayoutData.heightHint = 30;`
![image](https://user-images.githubusercontent.com/3328139/77846809-4b63e780-71d6-11ea-911a-d960c41825d7.png)

**After**: `linkLayoutData.heightHint = 40;` 
![image](https://user-images.githubusercontent.com/3328139/77846823-633b6b80-71d6-11ea-882e-1bd63f96c1a9.png)

tivervac replied:
LGTM

nbhusare commented:
@tivervac Tx for the code review. I have addressed your review comments and also added a few comments in reply concerning adding a safety check for `source` in the `AdvancedNewProjectPage.` I have made some minor changes to the `StatusWidget` as I observed some issues with it. Kindly have a look. Please resolve the conversations if you are ok with the changes made.

cdietrich commented:
thx @nbhusare
